Plutarch/PlutusType.hs

```diff
@@ -0,0 +1,335 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.PlutusType (
+  PlutusType (..),
+  PCon (..),
+  PMatch (..),
+) where
+
+import Data.Kind (Type)
+import Data.SOP.Constraint (AllZipF)
+import GHC.TypeLits (ErrorMessage (Text), Nat, TypeError, type (+))
+import Generics.SOP (
+  All,
+  AllZip,
+  AllZipN,
+  Generic (..),
+  I,
+  LiftedCoercible,
+  NP (..),
+  NS (..),
+  Prod,
+  SOP (SOP),
+  SameShapeAs,
+  Top,
+  hcoerce,
+  unSOP,
+ )
+import Plutarch.DataRepr.Internal.Generic (MkSum (mkSum))
+import Plutarch.DataRepr.Internal.HList (IndexList)
+import Plutarch.Internal (PType, S, Term, pforce, plam', punsafeCoerce, (:-->))
+import qualified Plutarch.Internal as PI
+import Plutarch.PLam ((#))
+
+{- |
+
+  The 'PlutusType' class allows encoding Haskell data-types as plutus terms
+  via constructors and destructors.
+
+  A simple example, encoding a Sum type as an Enum via PInteger:
+
+  > data AB (s :: S) = A | B
+  >
+  > instance PlutusType AB where
+  >   type PInner AB _ = PInteger
```

Comment:
We probably want to fix this documentation.

Reply:
I updated it to mention deriving for Scott encoding, assuming I didn't misinterpret what you meant here.